gk/fib.c

```diff
@@ -0,0 +1,1326 @@
+
+		/* TODO Add concurrency control. */
+		gt_fib->u.grantor.eth_cache->ref_cnt++;
+
+		goto out;
+	}
+
+	/* The Grantor server is an neighbor. */
+	if (gt_addr.proto == ETHER_TYPE_IPv4) {
+		eth_cache = lookup_ether_cache(
+			&gt_prefix_fib->u.neigh,
+			&gt_addr.ip.v4.s_addr);
+		if (eth_cache == NULL) {
+			ret = rte_hash_add_key(
+				gt_prefix_fib->u.neigh.hash_table,
+				&gt_addr.ip.v4.s_addr);
+			if (ret >= 0) {
+				eth_cache = &gt_prefix_fib->
+					u.neigh.cache_tbl[ret];
+			} else {
+				RTE_LOG(ERR, HASH,
+					"Failed to add a cache entry to the neighbor hash table at %s\n",
+					__func__);
+				rte_lpm_delete(ltbl->lpm,
+					ntohl(gt_addr.ip.v4.s_addr),
+					(sizeof(struct in_addr) * 8));
+				return NULL;
+			}
+		}
+
+	} else {
+		eth_cache = lookup_ether_cache(
+			&gt_prefix_fib->u.neigh6,
+			gt_addr.ip.v6.s6_addr);
+		if (eth_cache == NULL) {
+			ret = rte_hash_add_key(
+				gt_prefix_fib->u.neigh6.hash_table,
+				gt_addr.ip.v6.s6_addr);
+			if (ret >= 0) {
+				eth_cache = &gt_prefix_fib->
+					u.neigh6.cache_tbl[ret];
+			} else {
+				RTE_LOG(ERR, HASH,
+					"Failed to add a cache entry to the neighbor hash table at %s\n",
+					__func__);
+				rte_lpm6_delete(ltbl->lpm6,
+					gt_addr.ip.v6.s6_addr,
+					(sizeof(struct in6_addr) * 8));
+				return NULL;
+			}
+		}
+	}
+
+	eth_cache->stale = true;
+	eth_cache->eth_hdr.ether_type = gt_addr.proto;
+
+	/* Grantor server must be forwarded to the back network. `*/
+	ether_addr_copy(&gk_conf->net->back.eth_addr,
+		&eth_cache->eth_hdr.s_addr);
+
+	eth_cache->ref_cnt++;
+
+	gt_fib->u.grantor.eth_cache = eth_cache;
+	gt_fib->u.grantor.next_fib = gt_prefix_fib;
+	gt_fib->u.grantor.next_fib->ref_cnt++;
+
+out:
+	/* Initialize the fib entry for the IP prefix. */
+	ip_prefix_fib = add_prefix_fib(
+		&ip_prefix_addr, ip_prefix_len, gk_conf);
+	if (ip_prefix_fib != NULL) {
+		/* Fills up the FIB entry for the IP prefix. */
+		ip_prefix_fib->action = GK_FWD_GRANTOR;
+		rte_memcpy(&ip_prefix_fib->u,
+			&gt_fib->u, sizeof(ip_prefix_fib->u));
+
+		ip_prefix_fib->u.grantor.eth_cache->ref_cnt++;
+		gt_fib->u.grantor.next_fib->ref_cnt++;
+	} else {
+		del_fib_entry(grantor, gk_conf);
+		return NULL;
+	}
+
+	return gt_fib;
+}
+
+static struct gk_fib *
+init_drop_fib(const char *ip_prefix, struct gk_config *gk_conf)
+{
+	struct ipaddr ip_prefix_addr;
+	int ip_prefix_len;
+	struct gk_fib *ip_prefix_fib;
+
+	/* Initialize the fib entry for the IP prefix. */
+	ip_prefix_len = parse_ip_prefix(ip_prefix, &ip_prefix_addr);
+	if (ip_prefix_len < 0)
+		return NULL;
+
+	ip_prefix_fib = add_prefix_fib(
+		&ip_prefix_addr, ip_prefix_len, gk_conf);
+	if (ip_prefix_fib == NULL)
+		return NULL;
+
+	ip_prefix_fib->action = GK_DROP;
+
+	return ip_prefix_fib;
+}
+
+static int
+add_fib_entry(const char *ip_prefix, enum gk_fib_action action,
+	const char *grantor_or_gateway, struct gk_config *gk_conf)
+{
+	int ret;
+	struct gk_fib *gw_fib = NULL;
+	struct gk_fib *gt_fib = NULL;
```

Comment:
Yes, they will be further used once the TODOs are implemented. I will keep them local for now.